R/getExperimentCrossCorrelation.R

```diff
@@ -0,0 +1,577 @@
+#'
+#' @examples
+#' mae <- microbiomeDataSets::peerj32()
+#' 
+#' # Subset so that less observations / quicker to run, just for example
+#' mae[[1]] <- mae[[1]][1:20, 1:10]
+#' mae[[2]] <- mae[[2]][1:20, 1:10]
+#' # Calculate cross-correlations
+#' result <- getExperimentCrossCorrelation(mae, method = "pearson")
+#' # Show first 5 entries
+#' head(result, 5)
+#' 
+#' # Same can be done with TreeSummarizedExperiment and altExp
+#' # Create TreeSE with altExp
+#' tse <- mae[[1]]
+#' altExp(tse, "exp2") <- mae[[2]]
+#' result <- getExperimentCrossCorrelation(tse, experiment2 = "exp2", method = "pearson")
+#' # Show first 5 entries
+#' head(result, 5)
+#' 
+NULL
+
+#' @rdname getExperimentCrossCorrelation
+#' @export
+setGeneric("getExperimentCrossCorrelation", signature = c("x"),
+           function(x,
+                    experiment1 = 1,
+                    experiment2 = 2,
+                    abund_values1 = "counts",
+                    abund_values2 = "counts",
+                    method = "spearman",
+                    mode = "table",
+                    p_adj_method = "fdr",
+                    p_adj_threshold = 0.05,
+                    n_signif = 0,
+                    cor_threshold = NULL,
+                    sort = FALSE,
+                    filter_self_correlations = FALSE)
+             standardGeneric("getExperimentCrossCorrelation"))
+
+#' @rdname getExperimentCrossCorrelation
+#' @export
+setMethod("getExperimentCrossCorrelation", signature = c(x = "MultiAssayExperiment"),
+    function(x,
+             experiment1 = 1,
+             experiment2 = 2,
+             abund_values1 = "counts",
+             abund_values2 = "counts",
+             method = c("categorical", "pearson","spearman"),
+             mode = "table",
+             p_adj_method = "fdr",
+             p_adj_threshold = 0.05,
+             n_signif = 0,
+             cor_threshold = NULL,
+             sort = FALSE,
+             filter_self_correlations = FALSE){
+        ############################# INPUT CHECK ##############################
+        # Check experiment1 and experiment2
+        if( !( is.character(experiment1) && experiment1 %in% names(experiments(x)) || 
+            is.numeric(experiment1) && experiment1 <= length(experiments(x)) ) ){
+          stop("'experiment1' must be numeric or character value specifying", 
+               " experiment in experiment(x).", call. = FALSE)
+        }
+        if( !( is.character(experiment2) && experiment2 %in% names(experiments(x)) || 
+            is.numeric(experiment2) && experiment2 <= length(experiments(x)) ) ){
+          stop("'experiment2' must be numeric or character value specifying", 
+               " experiment in experiment(x).", call. = FALSE)
+        }
+        # If experiments are not the same, disable filter_self_correlations
+        if( experiment1 != experiment2 ){
+          filter_self_correlations <- FALSE
```

Comment:
So you mean because, we always use colwise-correlation in `getExperimentCrossCorrelation`, this would not be as clear for `getExperimentInterCorrelation`?

That could be an argument I guess. Maybe a name like `getSamplesInterCorrelation` would solve the issue.

Reply:
Correlations can be only calculated over samples because only there the tables are matched (same samples, different features). In principle, correlations over features could be done as well if the features were matched. But the real cases are possibly rare, so it is correlations between each pair of features, over their matched samples. 

Therefore I am not sure if if should be `getFeatureInterCorrelations` instead.

However only some of the association measures are correlations.

How about `getFeatureAssociations`? Do we need to specify if the associations are within one assay, or between assays? If the assay/altExp/MAE could be specified through the arguments, and one could also choose to do this with x vs. x (within a single assay) or x vs. y (inter correlations between different assays/altexp/mae)? It might be conceptuallly clear if we do not limit to the case where the two associated assays have to be different, user could also choose to do this for one assay against itself? Or does this add complexity rather than reduce it?